### PR TITLE
Trezor ui

### PR DESCRIPTION
**Pull request by adinael**
### Problem

The tpos contract is not supported using the trezor wallet

### Solution

Add a form to create a tpos contract using the trezor wallet 

### Result
Screenshot from 2020-10-08 10-00-25
Screenshot from 2020-08-27 23-52-52
Screenshot from 2020-10-08 09-55-56
Screenshot from 2020-10-08 10-17-42
Screenshot from 2020-10-08 10-18-17
Screenshot from 2020-10-08 10-18-51

### CODE REVIEW

File: web-ui/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (AlexITC) [inline web-ui/.vscode/launch.json]: I guess this isn't supposed to be committed, you could ignore the `.vscode/` directory.

File: web-ui/angular.json [hunk 2]
```diff
@@ -23,7 +23,10 @@
             "styles": [
               "src/styles.css"
             ],
-            "scripts": []
+            "scripts": [
```
Comment (AlexITC) [inline web-ui/angular.json]: Why are these scripts required?

File: web-ui/get-shims-on-ng-cli.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+const fs = require('fs');
```
Comment (AlexITC) [inline web-ui/get-shims-on-ng-cli.js]: This isn't used at all, is it?

File: web-ui/src/app/app.component.ts [hunk 4]
```diff
@@ -120,6 +120,13 @@ export class AppComponent implements OnInit {
       'label.normalFee': 'Normal',
       'label.lowFee': 'Low',
       'label.satoshis': 'Satoshis',
+      'label.merchantAddress': 'Merchant Address',
+      'label.merchantStakeAddress': 'MerchantStake address',
```
Comment (AlexITC) [inline web-ui/src/app/app.component.ts]: I guess this is the `Staking address`, which doesn't belong to the merchant but to the staker.

File: web-ui/src/app/components/trezor-connect/trezor-connect.component.html [hunk 5]
```diff
@@ -1,73 +1,157 @@
...
+            <select #selectedAddressType class="form-control" (change)="refresh()">
+              <option value=44>Legacy</option>
+              <option value=84>Segwit</option>
+              <option value=49 selected>P2shsegwit</option>
+            </select>
+          </div>
+          <div class="col-xs-3">
+            <button type="button" class="btn btn-primary" title="Add a new address"
+              (click)="generateNextAddress(+selectedAddressType.value)"><span class="glyphicon glyphicon-plus"></span>
+              Add</button>
```
Comment (AlexITC) [inline web-ui/src/app/components/trezor-connect/trezor-connect.component.html]: Missing i18n

File: web-ui/src/app/components/trezor-connect/trezor-connect.component.html [hunk 6]
```diff
@@ -1,73 +1,157 @@
...
+        <label class="first-position"> {{ 'label.createContract' | translate }}</label>
+
+        <form [formGroup]="tposContractFormControl" name="tposContractFormControl" (keyup.enter)="sendTPOS()" novalidate
+          autocomplete="off">
+          <div class="row">
+
+            <div class="form-group col-md-6 col-xs-8">
+              <label for="MerchantAddress">{{ 'label.merchantAddress' | translate }}</label>
+              <input type="text" class="form-control" aria-describedby="merchantAddress"
+                placeholder="Enter merchant address" formControlName="merchantAddress" name="merchantAddress">
```
Comment (AlexITC) [inline web-ui/src/app/components/trezor-connect/trezor-connect.component.html]: Missing i18n

File: web-ui/src/app/components/trezor-connect/trezor-connect.component.html [hunk 7]
```diff
@@ -1,73 +1,157 @@
...
+              <button type="button" class="btn btn-primary" [disabled]="tposContractFormControl.invalid"
+                (click)="sendTPOS()">{{ 'label.send' | translate }}</button>
+            </div>
+          </div>
+
+        </form>
+        <table class="table table-responsive table-borderless">
+          <thead>
+            <tr>
+              <th>Owner Address</th>
```
Comment (AlexITC) [inline web-ui/src/app/components/trezor-connect/trezor-connect.component.html]: Missing i18n

File: web-ui/src/app/components/trezor-connect/trezor-connect.component.ts [hunk 8]
```diff
@@ -32,20 +39,66 @@ export class TrezorConnectComponent implements OnInit {
...
+    private trezorRepositoryService: TrezorRepositoryService,
+    private tposContractsService: TposContractsService,
+    private notificationService: NotificationService
   ) { }
 
   ngOnInit() {
     TrezorConnect.manifest({
       email: 'trezor@wiringbits.net',
-      appUrl: 'https://xsnexplorer.io'
+      appUrl: 'http://localhost:9000'
```
Comment (AlexITC) [inline web-ui/src/app/components/trezor-connect/trezor-connect.component.ts]: You likely want to update this

File: web-ui/src/app/components/trezor-connect/trezor-connect.component.ts [hunk 9]
```diff
@@ -114,32 +169,129 @@ export class TrezorConnectComponent implements OnInit {
...
+    const collateral = convertToSatoshis(1); // The collateral is one xsn
+    const generatedInputs = this.generateInputs(collateral + fee);
+
+    if (generatedInputs.error) {
+      this.notificationService.error(generatedInputs.error);
+      return;
+    }
+
+    // create a new legacy address for contract
+    const tposAddress = await this.generateNextAddress(44);
```
Comment (AlexITC) [inline web-ui/src/app/components/trezor-connect/trezor-connect.component.ts:209]: Add a comment about the legacy address being required

File: web-ui/src/app/components/trezor-connect/trezor-connect.component.ts [hunk 10]
```diff
@@ -214,4 +363,61 @@ export class TrezorConnectComponent implements OnInit {
...
+    const raw = atob(str);
+    let result = '';
+    for (let i = 0; i < raw.length; i++) {
+      const hex = raw.charCodeAt(i).toString(16);
+      result += (hex.length === 2 ? hex : '0' + hex);
+    }
+    return result.toUpperCase();
+  }
+
+  isInt(n: any) {
```
Comment (AlexITC) [inline web-ui/src/app/components/trezor-connect/trezor-connect.component.ts]: This looks hacky, is there a better way?

**Review by AlexITC — APPROVED** [review]

File: web-ui/get-shims-on-ng-cli.js (hunk 3 above)
Comment (adinael) [reply web-ui/get-shims-on-ng-cli.js]: is not

File: web-ui/src/app/components/trezor-connect/trezor-connect.component.ts (hunk 10 above)
Comment (adinael) [reply web-ui/src/app/components/trezor-connect/trezor-connect.component.ts]: found a better way and fixed

**Pull request merged**